### PR TITLE
remove duplicated code and optimize

### LINKED ISSUE #21: Refactor repeated time check logic into a private function in Bootstrap.sol
In the Bootstrap.sol contract, the logic for checking spawn times and lock times is repeated multiple times. This logic can be refactored into a private function to reduce redundancy and improve maintainability. This issue tracks the suggested optimization for future implementation.

Related pull request: https://github.com/ExocoreNetwork/exocore-contracts/pull/17
Comment for context: https://github.com/ExocoreNetwork/exocore-contracts/pull/17#discussion_r1624704003

### LINKED ISSUE #68: Optimize using ++i to save cost in for loop instead of i++
It will save cost if using ++i, and initialize i in for loop.  Test report is below:
截屏2024-07-31 11 56 28
截屏2024-07-31 11 56 49

### PR DESCRIPTION
**Pull request by MaxMustermann2**
commit 3eb235ac65c5b38f2c5e9d52d21bc36f968df010
Author: MaxMustermann2 <82761650+MaxMustermann2@users.noreply.github.com>
Date:   Mon Aug 5 07:03:09 2024 +0000

    refactor: DRY time check
    
    Fixes #21

commit ca7d77caf0658ce97548db7df16435670f56be5a
Author: MaxMustermann2 <82761650+MaxMustermann2@users.noreply.github.com>
Date:   Mon Aug 5 06:53:17 2024 +0000

    fix: replace ++i with i++
    
    closes #68

commit 4af2fd6a2c4bea02ebdb42ebbc8cac71dfaed148
Author: MaxMustermann2 <82761650+MaxMustermann2@users.noreply.github.com>
Date:   Mon Aug 5 06:50:04 2024 +0000

    refactor(bootstrap): optimize checks
    
    Instead of iterating through the validator set, use a mapping to
    identify whether the consensus key and the name are unique.
    
    Closes #73 and #18

## Summary by CodeRabbit

- **New Features**
	- Enhanced validation for spawn time and offset duration in the Bootstrap contract.
	- Introduced checks for validator name length during registration.
	- Added mappings to track usage of consensus public keys and validator names.

- **Bug Fixes**
	- Improved error handling for empty validator names and zero consensus keys to prevent invalid registrations.

- **Tests**
	- Added new test cases for validator registration and updated assertions for key usage and spawn time verification.

### CODE REVIEW

File: src/core/Bootstrap.sol [hunk 1]
```diff
@@ -274,31 +256,18 @@ contract Bootstrap is
...
-        return false;
-    }
-
     /// @inheritdoc IValidatorRegistry
     function replaceKey(bytes32 newKey) external beforeLocked whenNotPaused {
         if (bytes(ethToExocoreAddress[msg.sender]).length == 0) {
             revert Errors.BootstrapValidatorNotExist();
         }
-        if (consensusPublicKeyInUse(newKey)) {
+        if (consensusPublicKeyInUse[newKey]) {
```
Comment (adu-web3) [inline src/core/Bootstrap.sol:264]: we should also check that consensusPublicKey is not zero?

File: src/core/Bootstrap.sol [hunk 2]
```diff
@@ -224,11 +223,11 @@ contract Bootstrap is
             revert Errors.BootstrapValidatorAlreadyRegistered();
         }
         // check that the consensus key is unique.
-        if (consensusPublicKeyInUse(consensusPublicKey)) {
+        if (consensusPublicKeyInUse[consensusPublicKey]) {
```
[the remark this answers is not in the document]
Comment (MaxMustermann2) [reply src/core/Bootstrap.sol:230]: Added

File: src/core/Bootstrap.sol (hunk 1 above)
Comment (MaxMustermann2) [reply src/core/Bootstrap.sol:273]: Added

**Review by adu-web3 — APPROVED** [review]
LGTM

**Review by magj2006 — APPROVED** [review]

**Pull request merged**